opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -208,201 +208,159 @@ void extendGIDtoRank(const GraphOfGrid<Dune::CpGrid>& gog,
 #if HAVE_MPI
 namespace Impl{
 
-std::vector<std::vector<std::vector<int>>>
+std::vector<std::vector<int>>
 extendRootExportList(const GraphOfGrid<Dune::CpGrid>& gog,
                      std::vector<std::tuple<int,int,char>>& exportList,
                      int root,
                      const std::vector<int>& gIDtoRank)
 {
     const auto& cc = gog.getGrid().comm();
     // non-root ranks have empty export lists.
-    std::vector<std::vector<std::vector<int>>> exportedWells;
+    std::vector<std::vector<int>> exportedCells;
     if (cc.rank()!=root)
     {
-        return exportedWells;
+        return exportedCells;
     }
-    exportedWells.resize(cc.size());
-    using ExportList = std::vector<std::tuple<int,int,char>>;
-    // make a list of wells for easy identification. Contains ID, begin, end
+
     using iter = std::set<int>::const_iterator;
-    std::unordered_map<int, std::tuple<iter,iter,int>> wellMap;
-    for (const auto& well : gog.getWells())
+    // store which wells are exported. Contains {begin, end, destination rank}
+    std::vector<std::tuple<iter,iter,int>> wellsToExport;
+    // track how many cells per rank are exported - to reserve the vector size
+    std::vector<int> sizesOfExport(cc.size(), 0);
+    const auto& gogWells = gog.getWells();
+    wellsToExport.reserve(gogWells.size());
+
+    // with gIDtoRank we can directly tell which wells belong where
+    if (gIDtoRank.size()>0)
     {
-        if (gIDtoRank.size()>0)
+        for (const auto& well : gogWells)
         {
             auto wellID = *well.begin();
             if (gIDtoRank[wellID]!=root)
             {
-                wellMap[wellID] = std::make_tuple(well.begin(), well.end(), well.size());
+                wellsToExport.emplace_back(well.begin(), well.end(), gIDtoRank[wellID]);
+                sizesOfExport[gIDtoRank[wellID]] += well.size()-1;
             }
         }
-        else
+    }
+    else
+    {
+        // make a list of wells for easy identification during search. Contains {begin, end, well ID}
+        std::unordered_map<int, std::tuple<iter,iter,int>> wellMap;
+        for (const auto& well : gogWells)
         {
             wellMap[*well.begin()] = std::make_tuple(well.begin(), well.end(), well.size());
         }
-    }
-
-    ExportList addToList;
-    // iterate once through the original exportList
-    for (const auto& cellProperties : exportList)
-    {
-        // if a cell is in any well, add cells of the well to exportList
-        auto pWell = wellMap.find(std::get<0>(cellProperties));
-        if (pWell!=wellMap.end())
+        // iterate once through the original exportList and identify exported wells
+        for (const auto& cellProperties : exportList)
         {
-            int rankToExport = std::get<1>(cellProperties);
-            if (rankToExport!=root)
+            auto pWell = wellMap.find(std::get<0>(cellProperties));
```

Comment:
That would work. The exportList even needs to be sorted when it is finished. (I actually removed the pre-sorting of the exportList in the last commit because it turned out to be unnecessary.)
However, .find() of unordered_map has complexity O(1) so the improvement would be minimal.

Reply:
> However, .find() of unordered_map has complexity O(1) so the improvement would be minimal.

I think the actual time spent depends highly on how the buckets of the undered map look like. Only looking up the bucket is fast. As we have not giving the unordered map  any information about the entries the buckets used might be far from optimal.

When using a map the operations are clear, not many and fast:
- check whether the current key is less than the cell
- Increment until it is the same of larger (this should be one increment)
- check wether the key is the cell

For an unordered map it is:
- calculate the bucket index
- Search in the bucket for the key
- check whether the key is the cell

So it really depends how often you hit a bucket with many entries

My guess would be that to compete with the map approach one would either need to have very few wells or very many buckets. In both cases this might come at the price of memory usage.

Reply:
For me, the main reason for using the `unordered_map` was the insertion and deletion of entries. We build the map, find each entry once, and delete entries after finding them.

Reply:
The benchmarks show that `map` is better than `unordered_map`, although the fastest option is skipping the search when gIDtoRank is available (default).
| Test case | gIDtoRank | map        | unordered_map |
| Easy         | 322              | 352         | 404                        |
| Medium  | 596841       | 654318   | 744448                 |
| Hard        | 5400064     | 5893489 | 6303968              |